### PR TITLE
lightstep: allow file-less access token option

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

This allows self-contained bootstrap without extra token files.

### CODE REVIEW

File: api/envoy/config/trace/v3/lightstep.proto [hunk 1]
```diff
@@ -40,9 +40,14 @@ message LightstepConfig {
   // The cluster manager cluster that hosts the LightStep collectors.
   string collector_cluster = 1 [(validate.rules).string = {min_len: 1}];
 
-  // File containing the access token to the `LightStep
-  // <https://lightstep.com/>`_ API.
-  string access_token_file = 2 [(validate.rules).string = {min_len: 1}];
+  oneof access_token_setting {
```
Comment (asraa) [inline api/envoy/config/trace/v3/lightstep.proto:43]: Will DataSource work?https://github.com/envoyproxy/envoy/blob/5a8bfa20dc3c85ecb61826d122696ecaa75dffa0/api/envoy/config/core/v3/base.proto#L334

https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/common/config/datasource.cc#L16

Comment (kyessenov) [reply api/envoy/config/trace/v3/lightstep.proto:43]: I suppose so. There would be an overlap with file data source, but that might be OK?

Comment (kyessenov) [reply api/envoy/config/trace/v3/lightstep.proto:45]: Somewhat tangential, is there a way to make a data source as confidential, so it's not leaked through config dump?

File: api/envoy/config/trace/v3/lightstep.proto [hunk 2]
```diff
@@ -40,9 +42,14 @@ message LightstepConfig {
...
```
Comment (lizan) [inline api/envoy/config/trace/v3/lightstep.proto:45]: You can't upgrade to oneof, it breaks the backward compatibility in generated code (e.g. Golang), which isn't allowed in our API policy. We should deprecate `access_token_file` and use DataSource.

**Review by lizan — CHANGES_REQUESTED** [review]

File: api/envoy/config/trace/v3/lightstep.proto (hunk 2 above)
Comment (kyessenov) [reply api/envoy/config/trace/v3/lightstep.proto]: OK, done.

**Review by lizan — APPROVED** [review]

**Review by lizan — DISMISSED** [review]

**Review by lizan — APPROVED** [review]

**Comment by moderation** [comment]
@kyessenov This PR broke file based token config for Lightstep for me.  `/stats/` would show spans being sent but using debug I can see that the requests generate `400` responses. No spans were being displayed on the Lightstep dashboard:
```
[2021-05-07 14:47:40.751][424004][debug][router] [source/common/router/router.cc:445] [C0][S12133036106175096015] cluster 'lightstep' match for URL '/lightstep.collector.CollectorService/Report'
[2021-05-07 14:47:40.751][424004][debug][router] [source/common/router/router.cc:631] [C0][S12133036106175096015] router decoding headers:                                                        
':method', 'POST'                                                                                                                                                                                 
':path', '/lightstep.collector.CollectorService/Report'                                                                                                                                           
':authority', 'lightstep'                                                                                                                                                                         
':scheme', 'http'                                                                                                                                                                                 
'te', 'trailers'                                                                                                                                                                                  
'grpc-timeout', '5000m'                                                                                                                                                                           
'content-type', 'application/grpc'                                                                                                                                                                
'x-envoy-internal', 'true'                                                                                                                                                                        
'x-forwarded-for', '192.168.7.4'                                                                                                                                                                  
'x-envoy-expected-rq-timeout-ms', '5000'                                                                                                                                                          
                                                                                                                                                                                                  
[2021-05-07 14:47:40.751][424004][debug][pool] [source/common/conn_pool/conn_pool_base.cc:232] [C4] using existing connection                                                                     
[2021-05-07 14:47:40.751][424004][debug][pool] [source/common/conn_pool/conn_pool_base.cc:171] [C4] creating stream                                                                               
[2021-05-07 14:47:40.751][424004][debug][router] [source/common/router/upstream_request.cc:371] [C0][S12133036106175096015] pool ready                                                            
[2021-05-07 14:47:40.856][424004][debug][client] [source/common/http/codec_client.cc:130] [C4] response complete                                                                                  
[2021-05-07 14:47:40.856][424004][debug][pool] [source/common/conn_pool/conn_pool_base.cc:200] [C4] destroying stream: 1 remaining                                                                
[2021-05-07 14:47:40.856][424004][debug][router] [source/common/router/router.cc:1229] [C0][S10426934778521263354] upstream headers complete: end_stream=true                                     
[2021-05-07 14:47:40.856][424004][debug][http] [source/common/http/async_client_impl.cc:104] async http request response headers (end_stream=true):                                               
':status', '400'                                                                                                                                                                                  
'date', 'Fri, 07 May 2021 21:47:40 GMT'                                                                                                                                                           
'server', 'envoy'                                                                                                                                                                                 
'x-envoy-upstream-service-time', '105'                                                                                                                                                            
```
with this config:
```json
                  "tracing": {
                    "provider": {
                      "name": "envoy.tracers.lightstep",
                      "typed_config": {
                        "@type": "type.googleapis.com/envoy.config.trace.v3.LightstepConfig",
                        "access_token": {
                          "filename": "lightstep_access_token"
                        },
                        "collector_cluster": "lightstep",
                        "propagation_modes": [
                          "LIGHTSTEP"
                        ]
                      }
                    }
                  }
```
Switching the config to an `inline_string` and it works.

**Comment by kyessenov** [comment]
@moderation can you try with fully qualified file name? I think I set it to allow empty file permitted, so the file not found error might be silent.

**Comment by kyessenov** [comment]
@moderation I suspect the difference is `rtrim` used in the old method. Does it happen to have some whitespace at the end?

**Comment by moderation** [comment]
Yes - that was it. There was a line break. I remove that and it works. Thanks. Not sure if we want to spell that out in the docs? I also noticed that `access_token_file` is not marked as deprecated at https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/lightstep.proto.html?highlight=lightstep

**Comment by kyessenov** [comment]
I think I did mark it deprecated, probably not refreshed yet. I'll just send a patch to trim right space, this might bite many people.

**Pull request merged**